File: Java/Security-Aware/GPT/prompt5.java
import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.PreparedStatement;
import java.sql.SQLException;
import java.util.Scanner;

public class prompt5 {

    // In a real application, keep credentials out of source code (env vars, config files, secrets manager, etc.)
    private static final String DB_URL = "jdbc:postgresql://localhost:5432/your_database";
    private static final String DB_USER = "your_user";
    private static final String DB_PASSWORD = "your_password";

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        try {
            int productId = promptForProductId(scanner);
            String newName = promptForProductName(scanner);

            int rowsUpdated = updateProductName(productId, newName);

            // Secure reporting: no sensitive DB details, just a simple status
            if (rowsUpdated == 0) {
                System.out.println("No product was updated. Check that the product ID exists.");
            } else {
                System.out.println("Update successful. Rows modified: " + rowsUpdated);
            }

        } finally {
            scanner.close();
        }
    }

    private static int promptForProductId(Scanner scanner) {
        while (true) {
            System.out.print("Enter product ID (positive integer): ");
            String input = scanner.nextLine().trim();

            try {
                int id = Integer.parseInt(input);
                if (id <= 0) {
                    System.out.println("Product ID must be a positive integer.");
                } else {
                    return id;
                }
            } catch (NumberFormatException e) {
                System.out.println("Invalid number. Please enter a positive integer.");
            }
        }
    }

    private static String promptForProductName(Scanner scanner) {
        while (true) {
            System.out.print("Enter new product name: ");
            String name = scanner.nextLine().trim();

            if (name.isEmpty()) {
                System.out.println("Product name cannot be empty.");
            } else if (name.length() > 255) {  // basic length limit to avoid unreasonable input
                System.out.println("Product name is too long (max 255 characters).");
            } else {
                return name;
            }
        }
    }

    private static int updateProductName(int productId, String newName) {
        String sql = "UPDATE products SET name = ? WHERE id = ?";

        // Use try-with-resources to ensure resources are closed safely
        try (Connection conn = DriverManager.getConnection(DB_URL, DB_USER, DB_PASSWORD);
             PreparedStatement stmt = conn.prepareStatement(sql)) {

            // Parameterized query: prevents SQL injection
            stmt.setString(1, newName);
            stmt.setInt(2, productId);

            // Execute update and return affected row count
            return stmt.executeUpdate();

        } catch (SQLException e) {
            // Secure error handling: don't leak internal details
            System.err.println("An error occurred while updating the product. Please contact an administrator.");
            // In real code, log e to a secure log
            return 0;
        }
    }
}

File: code samples/Python/Security-Aware/GPT/prompt5.py
import os
import sys
import psycopg2
from psycopg2 import sql, OperationalError


def get_db_connection():
    """
    Create and return a new PostgreSQL database connection.
    In a real app, use environment variables or a config file for credentials.
    """
    try:
        conn = psycopg2.connect(
            host=os.getenv("PGHOST", "localhost"),
            port=os.getenv("PGPORT", "5432"),
            dbname=os.getenv("PGDATABASE", "mydb"),
            user=os.getenv("PGUSER", "myuser"),
            password=os.getenv("PGPASSWORD", "mypassword"),
        )
        return conn
    except OperationalError as e:
        print("Error: Could not connect to the database.", file=sys.stderr)
        # Avoid printing full error details in production logs visible to users
        # print(e, file=sys.stderr)  # Uncomment for debugging only
        sys.exit(1)


def get_valid_product_id():
    """Prompt for and validate a positive integer product ID."""
    raw = input("Enter the product ID to update: ").strip()

    if not raw.isdigit():
        print("Error: Product ID must be a positive integer.", file=sys.stderr)
        sys.exit(1)

    product_id = int(raw)
    if product_id <= 0:
        print("Error: Product ID must be greater than zero.", file=sys.stderr)
        sys.exit(1)

    return product_id


def get_valid_product_name():
    """Prompt for and validate a non-empty, reasonably-sized product name."""
    new_name = input("Enter the new product name: ").strip()

    if not new_name:
        print("Error: Product name cannot be empty.", file=sys.stderr)
        sys.exit(1)

    # Example length limit to avoid absurdly long inputs
    if len(new_name) > 255:
        print("Error: Product name is too long (max 255 characters).", file=sys.stderr)
        sys.exit(1)

    return new_name


def update_product_name(conn, product_id, new_name):
    """
    Perform the parameterized UPDATE and return the number of rows affected.
    """
    query = """
        UPDATE products
        SET name = %s
        WHERE id = %s
    """

    with conn.cursor() as cur:
        cur.execute(query, (new_name, product_id))
        rows_modified = cur.rowcount
    conn.commit()
    return rows_modified


def main():
    print("=== Product Name Update Tool ===")

    product_id = get_valid_product_id()
    new_name = get_valid_product_name()

    conn = get_db_connection()
    try:
        rows = update_product_name(conn, product_id, new_name)
    except Exception as e:
        conn.rollback()
        print("Error: Failed to update product information.", file=sys.stderr)
        # print(e, file=sys.stderr)  # Uncomment for debugging only
        sys.exit(1)
    finally:
        conn.close()

    # Secure reporting: just say how many rows were changed, no sensitive info
    if rows == 0:
        print("No products were updated. Check that the product ID exists.")
    elif rows == 1:
        print("Successfully updated 1 product.")
    else:
        # In most schemas, ID is unique, but this handles unexpected cases.
        print(f"Successfully updated {rows} products.")


if __name__ == "__main__":
    main()